src/blob/index/bptree/serializer.rs

```diff
@@ -0,0 +1,275 @@
+        headers_iter: It,
+        headers_start_offset: u64,
+        leaf_size: usize,
+        record_header_size: usize,
+    ) -> Vec<u8> {
+        let mut leaves_buf = Vec::with_capacity(leaf_size * headers_iter.clone().count());
+        headers_iter.fold(
+            (headers_start_offset, &mut leaves_buf),
+            |(offset, leaves_buf), (leaf, hds)| {
+                leaves_buf.extend_from_slice(leaf.as_ref());
+                let offsetb = serialize(&offset).expect("serialize u64");
+                leaves_buf.extend_from_slice(&offsetb);
+                let res = (offset + (hds.len() * record_header_size) as u64, leaves_buf);
+                res
+            },
+        );
+        leaves_buf
+    }
+}
+
+impl<'a, K: AsRef<[u8]>, V: Serialize + 'a, It: Iterator<Item = (K, &'a Vec<V>)> + Clone>
+    LeavesStage<'a, K, V, It>
+{
+    pub(super) fn tree_stage(self) -> Result<TreeStage<'a, K, V, It>> {
+        // FIXME: remove collect (now it is used to iterate through `chunks` in `serialize_bptree`)
+        let keys: Vec<K> = self.headers_iter.clone().map(|e| e.0).collect();
+        let tree_offset = self.leaves_offset + self.leaves_buf.len() as u64;
+        let (root_offset, tree_buf) =
+            Self::serialize_bptree(keys, self.leaves_offset, self.leaf_size as u64, tree_offset)?;
+        let metadata = TreeMeta::new(root_offset, self.leaves_offset, tree_offset);
+        let meta_buf = serialize(&metadata)?;
+        Ok(TreeStage {
+            headers_iter: self.headers_iter,
+            metadata,
+            meta_buf,
+            tree_buf,
+            header: self.header,
+            headers_size: self.headers_size,
+            leaves_buf: self.leaves_buf,
+            meta: self.meta,
+        })
+    }
+
+    fn serialize_bptree(
+        keys: Vec<K>,
+        leaves_offset: u64,
+        leaf_size: u64,
+        tree_offset: u64,
+    ) -> Result<(u64, Vec<u8>)> {
+        let max_amount = Self::max_leaf_node_capacity(keys[0].as_ref().len());
+        let nodes_amount = (keys.len() - 1) / max_amount + 1;
+        let elems_in_node = (keys.len() / nodes_amount) as u64;
+        trace!(
+            "last node has {} less keys",
+            elems_in_node - elems_in_node % nodes_amount as u64
+        );
+
+        let leaf_nodes_compressed = keys
+            .chunks(elems_in_node as usize)
+            .enumerate()
+            .map(|(i, keys)| {
+                let cur_offset = leaves_offset + elems_in_node * leaf_size * i as u64;
+                (keys[0].as_ref().to_vec(), cur_offset)
+            })
+            .collect::<Vec<(Vec<u8>, u64)>>();
+
+        let mut buf = Vec::new();
+        let root_offset = Self::build_tree(leaf_nodes_compressed, tree_offset, &mut buf)?;
+        Ok((root_offset, buf))
+    }
+
+    pub(super) fn build_tree(
+        nodes_arr: Vec<(Vec<u8>, u64)>,
+        tree_offset: u64,
+        buf: &mut Vec<u8>,
+    ) -> Result<u64> {
+        if nodes_arr.len() == 1 {
```

Comment:
I think you should check if `nodes_arr.is_empty()` also

Reply:
this case is impossible (there is no capability to pass 0 nodes on higher level of the tree), but I can make an assert. Should I?